### PR TITLE
feat: add object restore functionality

### LINKED ISSUE #654: Add support for object softdelete
Need to add the ability to softdelete / restore objects as well as add the appropriate fields (Soft Delete Time, Hard Delete Time).

### PR DESCRIPTION
**Pull request by ddelgrosso1**
Fixes #654 

> It's a good idea to open an issue first for discussion.

- [ ] Tests pass
- [ ] Appropriate changes to README are included in PR

### CODE REVIEW

**Comment by ddelgrosso1** [comment]
This does not add the ability to list/get soft deleted objects. Should it?

**Comment by ddelgrosso1** [comment]
Will add get / list operations as well. Looks easy enough.

File: testbench/database.py [hunk 1]
```diff
@@ -354,6 +400,45 @@ def do_update_object(
...
+        context=None,
+    ) -> T:
+        with self._resources_lock:
+            bucket_with_metadata = self.get_bucket(bucket_name, context)
+            if not bucket_with_metadata.metadata.HasField("soft_delete_policy"):
+                testbench.error.invalid("SoftDeletePolicyRequired", context)
+            bucket = self.__get_bucket_for_object(bucket_name, context)
+            blob = bucket.get("%s#%d" % (object_name, generation), None)
+            if blob is not None:
+                testbench.error.not_soft_deleted(context)
```
Comment (cojenco) [inline testbench/database.py:424]: Is this the case where "If the specified object in the request is not a soft-deleted object but is either live or noncurrent, the request fails with a 412"?

File: testbench/database.py [hunk 2]
```diff
@@ -208,6 +214,43 @@ def __del_live_generation(self, bucket_name, object_name, context):
...
+        def soft_delete_filter(blob):
+            delta = now - blob.metadata.soft_delete_time.ToDatetime()
+            return delta.total_seconds() < retention_duration
+
+        if self._soft_deleted_objects[bucket_key].get(object_name) is not None:
+            self._soft_deleted_objects[bucket_key][object_name] = list(
+                filter(
+                    soft_delete_filter,
+                    self._soft_deleted_objects[bucket_key][object_name],
+                )
```
Comment (cojenco) [inline testbench/database.py:240]: This is so clever. Thanks for adding this!

File: tests/test_grpc_server.py [hunk 3]
```diff
@@ -1396,6 +1396,46 @@ def test_object_write_conditional_overwrite(self):
...
+            storage_pb2.RestoreObjectRequest(
+                bucket="projects/_/buckets/sd-bucket-name",
+                object="object-to-restore",
+                generation=initial_generation,
+            ),
+            context,
+        )
+        context.abort.assert_not_called()
+        self.assertIsNotNone(response)
+        self.assertNotEqual(initial_generation, blob.metadata.generation)
```
Comment (cojenco) [inline tests/test_grpc_server.py:1437]: hmm I may be missing something... do we want to `assertNotEqual(initial_generation, response.generation)`

File: testbench/database.py [hunk 4]
```diff
@@ -354,6 +400,45 @@ def do_update_object(
...
+            )
+            if blob is not None:
+                blob.metadata.create_time.FromDatetime(
+                    datetime.datetime.now(datetime.timezone.utc)
+                )
+                blob.metadata.ClearField("soft_delete_time")
+                blob.metadata.metageneration = 1
+                blob.metadata.generation = blob.metadata.generation + 1
+                if bucket_with_metadata.metadata.autoclass.enabled is True:
+                    blob.metadata.storage_class = "STANDARD"
```
Comment (cojenco) [inline testbench/database.py:442]: Nice catch!

**Review by cojenco — APPROVED** [review]
LGTM, just have a few questions. Thanks for adding this!

File: testbench/database.py [hunk 5]
```diff
@@ -354,6 +406,44 @@ def do_update_object(
...
+                self.insert_object(bucket_name, blob, context, preconditions)
+
+            return blob
+
```
Comment (cojenco) [inline testbench/database.py:446]: Does the blob need to be removed from `self._soft_deleted_objects[bucket_key][object_name]` after it's restored?

File: testbench/database.py (hunk 1 above)
Comment (ddelgrosso1) [reply testbench/database.py:449]: Correct. This is the case where the object is not soft deleted.

File: tests/test_grpc_server.py (hunk 3 above)
Comment (ddelgrosso1) [reply tests/test_grpc_server.py:1437]: You are correct. Fixing.

File: testbench/database.py (hunk 5 above)
Comment (ddelgrosso1) [reply testbench/database.py:471]: Good catch. Fixing.

Comment (ddelgrosso1) [reply testbench/database.py:486]: Done.

File: tests/test_grpc_server.py (hunk 3 above)
Comment (ddelgrosso1) [reply tests/test_grpc_server.py]: Done.

File: testbench/database.py [hunk 6]
```diff
@@ -208,9 +217,65 @@ def __del_live_generation(self, bucket_name, object_name, context):
...
+        )
+        if blob is None:
+            return testbench.error.notfound(object_name, context)
+        return blob
+
+    def __get_all_soft_deleted_objects(self, bucket_name, context):
+        bucket_key = self.__bucket_key(bucket_name, context)
+        all_soft_deleted = []
+        for soft_deleted_list in self._soft_deleted_objects[bucket_key].values():
+            all_soft_deleted.extend(soft_deleted_list)
```
Comment (cojenco) [inline testbench/database.py:272]: nit: consider sorting the blobs by increasing generation number

**Review by cojenco — APPROVED** [review]

**Pull request merged**